### PR TITLE
2626 add rx scheduler junit rule

### PR DESCRIPTION
**Pull request by Baron-Severin**
## Checklist

- [X] Confirm the **acceptance criteria** is fully satisfied in the issue(s) this PR will close
- [X] Add thorough **tests** or an explanation of why it does not
- [X] Add a **CHANGELOG entry** if applicable
Not user facing
- [X] Add **QA labels** on the associated issue (not this PR; `qa-ready` or `qa-notneeded`)

### CODE REVIEW

File: app/src/test/java/org/mozilla/tv/firefox/helpers/RxSchedulerBeforeClass.kt [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+import org.junit.BeforeClass
+import org.junit.rules.ExternalResource
+import java.util.concurrent.Callable
+
+/**
+ * Used to force all tested Rx code to execute synchronously.
+ *
+ * IMPORTANT NOTE: **must** be called from [BeforeClass]
+ */
+fun forceRxSynchronous() {
```
Comment (mcomella) [inline app/src/test/java/org/mozilla/tv/firefox/helpers/RxSchedulerBeforeClass.kt]: nit: maybe call `beforeClassForceRxSynchronous` or `forceRxSynchronousInBeforeClass`?

File: app/src/test/java/org/mozilla/tv/firefox/helpers/RxSchedulerBeforeClass.kt [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+/**
+ * Used to force all tested Rx code that changes threads to operate on the same
+ * [TestScheduler].
+ *
+ * IMPORTANT NOTE: **must** be called from [BeforeClass]
+ *
+ * @return the [TestScheduler] to which all observeOn and subscribeOn calls will
+ * be forwarded.
+ */
+fun forceRxTestScheduler(): TestScheduler {
```
Comment (mcomella) [inline app/src/test/java/org/mozilla/tv/firefox/helpers/RxSchedulerBeforeClass.kt]: nit: same

File: app/src/test/java/org/mozilla/tv/firefox/helpers/RxSchedulerBeforeClass.kt (hunk 1 above)
Comment (mcomella) [inline app/src/test/java/org/mozilla/tv/firefox/helpers/RxSchedulerBeforeClass.kt]: nit: why are these declared at the top-level? In general, I'd prefer to namespace things inside a class or object.

File: app/src/test/java/org/mozilla/tv/firefox/helpers/RxSchedulerBeforeClass.kt [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    val testScheduler = TestScheduler()
+    setRxScheduler(testScheduler)
+    return testScheduler
+}
+
+private fun setRxScheduler(scheduleTo: Scheduler) {
+    val initHandler = Function<Callable<Scheduler>, Scheduler> { scheduleTo }
+    val setHandler = Function<Scheduler, Scheduler> { scheduleTo }
+
+    RxJavaPlugins.setInitIoSchedulerHandler(initHandler)
```
Comment (mcomella) [inline app/src/test/java/org/mozilla/tv/firefox/helpers/RxSchedulerBeforeClass.kt]: In the original code, we only used two of these: why are all of these necessary?

File: app/src/test/java/org/mozilla/tv/firefox/helpers/RxSchedulerBeforeClass.kt [hunk 4]
```diff
@@ -11,12 +11,23 @@ import io.reactivex.plugins.RxJavaPlugins
 import io.reactivex.schedulers.Schedulers
```
Comment (mcomella) [inline app/src/test/java/org/mozilla/tv/firefox/helpers/RxSchedulerBeforeClass.kt]: nit: this isn't a `@Rule` so we shouldn't call it that in the commit message.

File: app/src/test/java/org/mozilla/tv/firefox/helpers/tests/ForceRxSynchronousTest.kt [hunk 5]
```diff
@@ -0,0 +1,57 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
```
Comment (mcomella) [inline app/src/test/java/org/mozilla/tv/firefox/helpers/tests/ForceRxSynchronousTest.kt]: nit: these also aren't rules: they're just functions.

File: app/src/test/java/org/mozilla/tv/firefox/helpers/tests/ForceRxSynchronousTest.kt [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+        println(nameSet)
+
+        assertEquals(1, nameSet.size)
+    }
+}
+
+private fun <T> Observable<T>.toThreadName(): String = this
+    .map { Thread.currentThread().name }
+    .blockingFirst()
```
Comment (mcomella) [inline app/src/test/java/org/mozilla/tv/firefox/helpers/tests/ForceRxSynchronousTest.kt]: nit: nl at end of file

File: app/src/test/java/org/mozilla/tv/firefox/helpers/tests/ForceRxSynchronousTest.kt [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+
+        val newThread = Observable.just(1)
+            .observeOn(Schedulers.newThread())
+            .toThreadName()
+
+        val single = Observable.just(1)
+            .observeOn(Schedulers.single())
+            .toThreadName()
+
+        val nameSet = setOf(comp, io, trampoline, newThread, single)
```
Comment (mcomella) [inline app/src/test/java/org/mozilla/tv/firefox/helpers/tests/ForceRxSynchronousTest.kt]: nit: this decouples the definition of the Observerables to the definition of the set and leaves room for error when this code is changed. I'd prefer something like:
```kotlin
val nameSet = listOf(
    Schedulers.computation(),
    Schedulers.io(),
    ...
).map { Observable.just(1).observeOn(it).toThreadName() }
```

File: app/src/test/java/org/mozilla/tv/firefox/helpers/RxSchedulerBeforeClass.kt [hunk 8]
```diff
@@ -12,10 +12,16 @@ import io.reactivex.schedulers.Schedulers
...
 import java.util.concurrent.Callable
+import org.junit.ClassRule
 
 /**
  * Used to force all tested Rx code to execute synchronously.
  *
+ * IMPORTANT NOTE: **must** be called from [BeforeClass]
+ *
+ * This cannot be a JUnit [ClassRule] because [ClassRule]s are not executed in any
+ * files using Robolectric
```
Comment (mcomella) [inline app/src/test/java/org/mozilla/tv/firefox/helpers/RxSchedulerBeforeClass.kt]: nit: link to the issue

File: app/src/test/java/org/mozilla/tv/firefox/helpers/RxSchedulerBeforeClass.kt [hunk 9]
```diff
@@ -40,6 +44,9 @@ fun forceRxSynchronous() {
  *
  * IMPORTANT NOTE: **must** be called from [BeforeClass]
  *
+ * This cannot be a JUnit [ClassRule] because [ClassRule]s are not executed in any
+ * files using Robolectric
```
Comment (mcomella) [inline app/src/test/java/org/mozilla/tv/firefox/helpers/RxSchedulerBeforeClass.kt]: nit: link to issue

File: app/src/test/java/org/mozilla/tv/firefox/helpers/RxSchedulerBeforeClass.kt (hunk 1 above)
Comment (Baron-Severin) [reply app/src/test/java/org/mozilla/tv/firefox/helpers/RxSchedulerBeforeClass.kt]: My bad, I always forget that.

File: app/src/test/java/org/mozilla/tv/firefox/helpers/RxSchedulerBeforeClass.kt (hunk 3 above)
Comment (Baron-Severin) [reply app/src/test/java/org/mozilla/tv/firefox/helpers/RxSchedulerBeforeClass.kt]: `CursorModelTest` only uses the Android main thread and the computation scheduler, but other code could use all the rest of these.  I've personally only ever wanted all of my schedulers to operate the same way during tests, and if we ever run into a different use case we can always just not use this.

**Pull request merged**